components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/UserEmailVerificationHandler.java

```diff
@@ -272,5 +334,145 @@ protected User getUser(Map eventProperties, UserStoreManager userStoreManager){
+                }
+
+                if (emailAddress.equals(existingEmail)) {
+                    if (log.isDebugEnabled()) {
+                        log.debug(String.format("The email address to be updated is same as the existing email " +
+                                "address for user: %s . Hence an email notification will not be triggered.", username));
+                    }
+                    return;
+                }
+
+                claims.put(IdentityRecoveryConstants.VERIFICATION_PENDING_EMAIL_CLAIM, emailAddress);
+                claims.remove(IdentityRecoveryConstants.EMAIL_ADDRESS_CLAIM);
+            }
+        }
+    }
+
+    private void postSetUserClaimsOnEmailUpdate(User user, UserStoreManager userStoreManager) throws
+            IdentityEventException {
+
+        boolean isNotificationInternallyManage = Boolean.parseBoolean(Utils.getConnectorConfig
+                (IdentityRecoveryConstants.ConnectorConfig.EMAIL_VERIFICATION_NOTIFICATION_INTERNALLY_MANAGE,
+                        user.getTenantDomain()));
+
+        Claim verifyEmailClaim = Utils.getEmailVerifyTemporaryClaim();
+        if (verifyEmailClaim == null) {
+            // Not required to handle in this handler.
+            return;
+        }
+
+        String pendingVerificationEmailClaimValue = getPendingVerificationEmailValue(userStoreManager, user);
+        boolean isAccountClaimExist = Utils.isAccountStateClaimExisting(user.getTenantDomain());
+
+        if (IdentityRecoveryConstants.VERIFY_EMAIL_CLIAM.equals(verifyEmailClaim.getClaimUri()) &&
+                Boolean.parseBoolean(verifyEmailClaim.getValue()) && StringUtils.isNotBlank
+                (pendingVerificationEmailClaimValue)) {
+
+            if (isNotificationInternallyManage) {
+                if (isAccountClaimExist) {
+                    setUserClaim(IdentityRecoveryConstants.ACCOUNT_STATE_CLAIM_URI,
+                            IdentityRecoveryConstants.PENDING_EMAIL_VERIFICATION, userStoreManager, user);
+                }
+                initNotificationForEmailVerificationOnUpdate(pendingVerificationEmailClaimValue, user);
+            }
+        }
+    }
+
+    /**
+     * Checks whether the verificationPendingEmail claim exists and returns true if the claim exists, else returns
+     * false.
+     *
+     * @param tenantDomain Tenant Domain
+     * @return true if verificationPendingEmail claim exists else return false.
+     * @throws IdentityEventException IdentityEventException
+     */
+    private boolean isVerificationPendingEmailClaimConfigAvailable(String tenantDomain) throws IdentityEventException {
+
+        UserRealm userRealm = null;
+        ClaimManager claimManager = null;
+        boolean isExist = false;
+
+        RealmService realmService = IdentityRecoveryServiceDataHolder.getInstance().getRealmService();
+        if (realmService != null) {
```

Comment:
Don't we have to add a debug log saying that we are returning `false` since `realmService` or `userRealm` is `NULL`?